src/treq/test/test_treq_integration.py

```diff
@@ -241,6 +243,156 @@ def test_failed_basic_auth(self):
         self.assertEqual(response.code, 401)
         yield print_response(response)
 
+    @inlineCallbacks
+    def test_digest_auth(self):
+        """
+            Test successful Digest authentication
+        :return:
+        """
+        response = yield self.get('/digest-auth/auth/treq/treq',
+                                  auth=HTTPDigestAuth('treq', 'treq'))
+        self.assertEqual(response.code, 200)
+        yield print_response(response)
+        json = yield treq.json_content(response)
+        self.assertTrue(json['authenticated'])
+        self.assertEqual(json['user'], 'treq')
+
+    @inlineCallbacks
+    def test_digest_auth_multi_qop(self):
+        """
+            Test successful Digest authentication with multiple qop types
+        :return:
+        """
+        response = yield self.get('/digest-auth/undefined/treq/treq',
+                                  auth=HTTPDigestAuth('treq', 'treq'))
+        self.assertEqual(response.code, 200)
+        yield print_response(response)
+        json = yield treq.json_content(response)
+        self.assertTrue(json['authenticated'])
+        self.assertEqual(json['user'], 'treq')
+
+    @inlineCallbacks
+    def test_digest_auth_multiple_calls(self):
+        """
+            Test proper Digest authentication credentials caching
+        """
+
+        # A mutable holder for call counter
+        agent_request_call_storage = {
+            'c': 0,
+            'i': []
+        }
+
+        # Original Agent request call
+        agent_request_orig = Agent.request
+
+        def agent_request_patched(*args, **kwargs):
+            """
+                Patched Agent.request function,
+                that inscreaces call count on every HTTP request
+                and appends
+            """
+            response_deferred = agent_request_orig(*args, **kwargs)
+            agent_request_call_storage['c'] += 1
+            agent_request_call_storage['i'].append((args, kwargs))
+            return response_deferred
+
+        patcher = MonkeyPatcher(
+            (Agent, 'request', agent_request_patched)
+        )
+        patcher.patch()
+
+        auth = HTTPDigestAuth('treq-digest-auth-multiple', 'treq')
+
+        response1 = yield self.get(
+            '/digest-auth/auth/treq-digest-auth-multiple/treq',
+            auth=auth
+        )
+        self.assertEqual(response1.code, 200)
+        yield print_response(response1)
+        json1 = yield treq.json_content(response1)
+
+        # Assume we did two actual HTTP requests - one to obtain credentials
+        # and second is original request with authentication
+        self.assertEqual(
+            agent_request_call_storage['c'],
+            2
+        )
+        headers_for_second_request = agent_request_call_storage['i'][1][0][3]
```

Comment:
I can live with the messy monkey patching but please let's capture the arguments by name so this deep indexing is intelligible.

Reply:
Refactored to remove deep indexing and made variable names clearer.